src/detect-metadata.c

```diff
@@ -28,23 +28,175 @@
 
 #include "suricata-common.h"
 #include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
 #include "detect-metadata.h"
 
+#include "util-unittest.h"
+
+#define PARSE_REGEX "^\\s*([^\\s]+)\\s+([^\\s]+)(?:,\\s*([^\\s]+)\\s+([^\\s]+))*$"
+#define PARSE_TAG_REGEX "\\s*([^\\s]+)\\s+([^,]+)\\s*"
+
+static pcre *parse_regex;
+static pcre_extra *parse_regex_study;
+static pcre *parse_tag_regex;
+static pcre_extra *parse_tag_regex_study;
+
 static int DetectMetadataSetup (DetectEngineCtx *, Signature *, const char *);
+static void DetectMetadataRegisterTests(void);
 
 void DetectMetadataRegister (void)
 {
     sigmatch_table[DETECT_METADATA].name = "metadata";
-    sigmatch_table[DETECT_METADATA].desc = "ignored by suricata";
+    sigmatch_table[DETECT_METADATA].desc = "partially used by suricata";
     sigmatch_table[DETECT_METADATA].url = DOC_URL DOC_VERSION "/rules/meta.html#metadata";
     sigmatch_table[DETECT_METADATA].Match = NULL;
     sigmatch_table[DETECT_METADATA].Setup = DetectMetadataSetup;
     sigmatch_table[DETECT_METADATA].Free  = NULL;
-    sigmatch_table[DETECT_METADATA].RegisterTests = NULL;
+    sigmatch_table[DETECT_METADATA].RegisterTests = DetectMetadataRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex, &parse_regex_study);
+    DetectSetupParseRegexes(PARSE_TAG_REGEX, &parse_tag_regex, &parse_tag_regex_study);
+}
+
+/**
+ *  \brief Free a Metadata object
+ */
+void DetectMetadataFree(DetectMetadata *mdata)
+{
+    SCEnter();
+
+    if (mdata->key != NULL) {
+        pcre_free_substring(mdata->key);
+    }
+    if (mdata->value != NULL) {
+        pcre_free_substring(mdata->value);
+    }
+    SCFree(mdata);
+
+    SCReturn;
 }
 
-static int DetectMetadataSetup (DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+static int DetectMetadataParse(Signature *s, const char *metadatastr)
 {
+#define MAX_SUBSTRINGS 30
+    int ret = 0, res = 0;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = pcre_exec(parse_regex, parse_regex_study, metadatastr,
+                    strlen(metadatastr), 0, 0, ov, MAX_SUBSTRINGS);
+    if (ret < 2) {
+        SCLogError(SC_ERR_PCRE_MATCH, "pcre_exec parse error, ret %" PRId32
```

Comment:
I still worry about suddenly parsing a field that we've declared 'ignored' for many years. How knows what ppl may have stuffed into it? So I think we should be liberal about what we encounter, so not error and not be noisy. This has it's own drawback obviously, so perhaps we can add a config option to ignore the contents of the metadata field.

The parsing error here could refer to that then.